stacks-signer/src/coordinator.rs

```diff
@@ -0,0 +1,230 @@
+    pub last_message_time: Option<Instant>,
+    /// The time the coordinator started its tenure
+    tenure_start: Instant,
+    /// The public keys of the coordinators
+    public_keys: PublicKeys,
+}
+
+impl From<PublicKeys> for CoordinatorSelector {
+    /// Create a new Coordinator selector from the given list of public keys
+    fn from(public_keys: PublicKeys) -> Self {
+        let coordinator_ids =
+            CoordinatorSelector::calculate_coordinator_ids(&public_keys, &ConsensusHash::empty());
+        let coordinator_id = *coordinator_ids
+            .first()
+            .expect("FATAL: No registered signers");
+        let coordinator_index = 0;
+        let last_message_time = None;
+        let tenure_start = Instant::now();
+        Self {
+            coordinator_ids,
+            coordinator_id,
+            coordinator_index,
+            last_message_time,
+            tenure_start,
+            public_keys,
+        }
+    }
+}
+
+impl CoordinatorSelector {
+    /// Update the coordinator id
+    fn update_coordinator(&mut self, new_coordinator_ids: Vec<u32>) {
+        self.last_message_time = None;
+        self.coordinator_index = if new_coordinator_ids != self.coordinator_ids {
+            // We have advanced our block height and should select from the new list
+            let mut new_index: usize = 0;
+            self.coordinator_ids = new_coordinator_ids;
+            let new_coordinator_id = *self
+                .coordinator_ids
+                .first()
+                .expect("FATAL: No registered signers");
+            if new_coordinator_id == self.coordinator_id {
+                // If the newly selected coordinator is the same as the current and we have more than one available, advance immediately to the next
+                if self.coordinator_ids.len() > 1 {
+                    new_index = new_index.saturating_add(1);
+                }
+            }
+            new_index
+        } else {
+            let mut new_index = self.coordinator_index.saturating_add(1);
+            if new_index == self.coordinator_ids.len() {
+                // We have exhausted all potential coordinators. Go back to the start
+                new_index = 0;
+            }
+            new_index
+        };
+        self.coordinator_id = *self
+            .coordinator_ids
+            .get(self.coordinator_index)
+            .expect("FATAL: Invalid number of registered signers");
+        self.tenure_start = Instant::now();
+        self.last_message_time = None;
+    }
+
+    /// Check the coordinator timeouts and update the selected coordinator accordingly
+    /// Returns the resulting coordinator ID. (Note: it may be unchanged)
+    pub fn refresh_coordinator(&mut self, pox_consensus_hash: &ConsensusHash) -> u32 {
+        let new_coordinator_ids =
+            Self::calculate_coordinator_ids(&self.public_keys, pox_consensus_hash);
+        if let Some(time) = self.last_message_time {
```

Comment:
There's going to be some skew in this algorithm, depending on how fast or slow each signer runs.  It's possible that at or around `COORDINATOR_OPERATION_TIMEOUT_SECS` or `COORDINATOR_TENURE_TIMEOUT_SECS` that some signers will have chosen a new coordinator while others have not.  How certain are we that these two timeouts are sufficiently large that all signers will agree on the same coordinator almost all of the time?

Reply:
I would strongly advocate that this PR not be considered an implementation of the coordinator selection and reselection algorithm. That is a very big can of worms, and I think can and therefore should be separated from a first implementation of a signer binary that can perform DKG, submit aggregate-key votes, and issue block signatures in response to miner proposals.

Reply:
Ha, sure thing

Reply:
I def was having a lot of issues with coordinator selection (ostly in testing) and this was def a temp solution to mitigate. Needs a major hard look post this merge.

Reply:
https://github.com/stacks-network/stacks-core/issues/4420